File: python/planout/namespace.py
from abc import ABCMeta, abstractmethod, abstractproperty
from operator import itemgetter

from .experiment import Experiment, DefaultExperiment
from .ops.random import Sample, RandomInteger
from .assignment import Assignment

# decorator for methods that assume assignments have been made
def requires_experiment(f):
  def wrapped_f(self, *args, **kwargs):
    if not self._experiment:
      self._assign_experiment()
    return f(self, *args, **kwargs)
  return wrapped_f


def requires_default_experiment(f):
  def wrapped_f(self, *args, **kwargs):
    if not self._default_experiment:
      self._assign_default_experiment()
    return f(self, *args, **kwargs)
  return wrapped_f


class Namespace(object):
  __metaclass__ = ABCMeta
  def __init__(self, **kwargs):
    pass

  @abstractmethod
  def add_experiment(self, name, exp_object, num_segments, **kwargs):
    pass

  @abstractmethod
  def remove_experiment(self, name):
    pass

  @abstractmethod
  def set_auto_exposure_logging(self, value):
    pass

  @abstractproperty
  def in_experiment(self):
    pass

  @abstractmethod
  def get(self, name, default):
    pass

  @abstractmethod
  def log_exposure(self, extras=None):
    pass

  @abstractmethod
  def log_event(self, event_type, extras=None):
    pass

class SimpleNamespace(Namespace):
  __metaclass__ = ABCMeta
  def __init__(self, **kwargs):
    self.name = self.__class__  # default name is the class name
    self.inputs = kwargs
    self.num_segments = None

    # dictionary mapping segments to experiment names
    self.segment_allocations = {}

    # dictionary mapping experiment names to experiment objects
    self.current_experiments = {}

    self._experiment = None          # memoized experiment object
    self._default_experiment = None  # memoized default experiment object
    self.default_experiment_class = DefaultExperiment

    # setup name, primary key, number of segments, etc
    self.setup()
    self.available_segments = set(range(self.num_segments))

    # load namespace with experiments
    self.setup_experiments()


  @abstractmethod
  def setup(self):
    """Sets up experiment"""
    # Developers extending this class should set the following variables
    # self.name = 'sample namespace'
    # self.primary_unit = 'userid'
    # self.num_segments = 10000
    pass

  @abstractmethod
  def setup_experiments():
    # e.g.,
    # self.add_experiment('first experiment', Exp1, 100)
    pass

  @property
  def primary_unit(self):
    return self._primary_unit

  @primary_unit.setter
  def primary_unit(self, value):
    # later on we require that the primary key is a list, so we use
    # a setter to convert strings to a single element list
    if type(value) is list:
      self._primary_unit = value
    else:
      self._primary_unit = [value]


  def add_experiment(self, name, exp_object, segments):
    num_avail = len(self.available_segments)
    if num_avail < segments:
      print 'error: %s segments requested, only %s available.' % \
        (segments, num_avail)
      return False
    if name in self.current_experiments:
      print 'error: there is already an experiment called %s.' %  name
      return False

    # randomly select the given number of segments from all available segments
    a = Assignment(self.name)
    a.sampled_segments = \
      Sample(choices=list(self.available_segments), draws=segments, unit=name)

    # assign each segment to the experiment name
    for segment in a.sampled_segments:
      self.segment_allocations[segment] = name
      self.available_segments.remove(segment)

    # associate the experiment name with an object
    self.current_experiments[name] = exp_object


  def remove_experiment(self, name):
    if name not in self.current_experiments:
      print 'error: there is no experiment called %s.' %  name
      return False

    segments_to_free = \
      [s for s, n in self.segment_allocations.iteritems() if n == name]

    for segment in segments_to_free:
      del self.segment_allocations[segment]
      self.available_segments.add(segment)
    del self.current_experiments[name]

    return True

  def get_segment(self):
    # randomly assign primary unit to a segment
    a = Assignment(self.name)
    a.segment = RandomInteger(min=0, max=self.num_segments,
      unit=itemgetter(*self.primary_unit)(self.inputs))
    return a.segment


  def _assign_experiment(self):
    "assign primary unit to an experiment"

    segment = self.get_segment()
    # is the unit allocated to an experiment?
    if segment in self.segment_allocations:
      experiment_name = self.segment_allocations[segment]
      experiment = self.current_experiments[experiment_name](**self.inputs)
      experiment.name = '%s-%s' % (self.name, experiment_name)
      experiment.salt = '%s.%s' % (self.name, experiment_name)
      self._experiment = experiment
      self._in_experiment = experiment.in_experiment
    else:
      self._assign_default_experiment()
      self._in_experiment = False


  def _assign_default_experiment(self):
    self._default_experiment = self.default_experiment_class(**self.inputs)


  @requires_default_experiment
  def default_get(self, name, default=None):
    return self._default_experiment.get(name, default)


  @property
  @requires_experiment
  def in_experiment(self):
    return self._in_experiment


  @in_experiment.setter
  def in_experiment(self, value):
    # in_experiment cannot be externally modified
    pass


  @requires_experiment
  def set_auto_exposure_logging(self, value):
    self._experiment.set_auto_exposure_logging(value)

  @requires_experiment
  def get(self, name, default=None):
    if self._experiment is None:
      return self.default_get(name, default)
    else:
      return self._experiment.get(name, self.default_get(name, default))

  @requires_experiment
  def log_exposure(self, extras=None):
    """Logs exposure to treatment"""
    if self._experiment is None:
      pass
    self._experiment.log_exposure(extras)

  @requires_experiment
  def log_event(self, event_type, extras=None):
    """Log an arbitrary event"""
    if self._experiment is None:
      pass
    self._experiment.log_event(event_type, extras)

File: alpha/js/es6/namespace.js
import Experiment from "./experiment.js";
import Assignment from "./assignment.js";
import { Sample, RandomInteger } from "./ops/random.js";
import { range, isObject, forEach } from "./lib/utils.js";


class DefaultExperiment extends Experiment {
  configure_logger() {
    return;
  }

  setup() {
    this.name = 'test_name';
  }

  log(data) {
    return;
  }

  previously_logged() {
    return true;
  }

  assign(params, args) {
    return;
  }
}

class Namespace {

  add_experiment(name, obj, segments) {
    throw "IMPLEMENT add_experiment";
  }

  remove_experiment(name) {
    throw "IMPLEMENT remove_experiment";
  }

  set_auto_exposure_logging(value) {
    throw "IMPLEMENT set_auto_exposure_logging";
  }

  in_experiment() {
    throw "IMPLEMENT in_experiment";
  }

  get(name, default_val) {
    throw "IMPLEMENT get";
  }

  log_exposure(extras) {
    throw "IMPLEMENT log_exposure";
  }

  log_event(event_type, extras) {
    throw "IMPLEMENT log_event";
  }

  require_experiment() {
    if (!this._experiment) {
      this._assign_experiment();
    }
  }

  require_default_experiment() {
    if (!this._default_experiment) {
      this._assign_default_experiment();
    }
  }
}

class SimpleNamespace extends Namespace {
  
  constructor(args) {
    super(args);
    this.name = this.getDefaultNamespaceName();
    this.inputs = args;
    this.num_segments = 1;
    this.segment_allocations = {};
    this.current_experiments = {};

    this._experiment = null;
    this._default_experiment = null;
    this.default_experiment_class = DefaultExperiment
    this._in_experiment = false;

    this.setup();
    this.available_segments = range(this.num_segments);

    this.setup_experiments();
  }

  setup() {
    throw "IMPLEMENT setup";
  }

  setup_experiments() {
    throw "IMPLEMENT setup_experiments";
  }

  get_primary_unit() {
    return this._primary_unit;
  }

  set_primary_unit(value) {
    this._primary_unit = value;
  }

  add_experiment(name, exp_object, segments) {
    var number_available = this.available_segments.length;
    if (number_available < segments) {
      return false;
    } else if (this.current_experiments[name] !== undefined) {
      return false;
    }
    var a = new Assignment(this.name);
    a.set('sampled_segments', new Sample({'choices': this.available_segments, 'draws': segments, 'unit': name}));
    var sample = a.get('sampled_segments');
    for(var i = 0; i < sample.length; i++) {
      this.segment_allocations[sample[i]] = name;
      this.available_segments.splice(this.available_segments.indexOf(sample[i]), 1);
    }
    this.current_experiments[name] = exp_object
    
  }

  remove_experiment(name) {
    if (this.current_experiments[name] === undefined) {
      return false;
    }

    var segments_to_free = [];
    forEach(Object.keys(this.segment_allocations), (cur) => {
      if(this.segment_allocations[cur] === name) {
        segments_to_free.push(cur);
      }
    });
    for (var i = 0; i < segments_to_free.length; i++) {
      var segment = segments_to_free[i];
      delete this.segment_allocations[segment];
      this.available_segments.push(segment);
    }
    delete this.current_experiments[name];
    return true;
  }

  get_segment() {
    var a = new Assignment(this.name);
    var segment = new RandomInteger({'min': 0, 'max': this.num_segments-1, 'unit': this.inputs[this.get_primary_unit()]});
    a.set('segment', segment);
    return a.get('segment');
  }

  _assign_experiment() {
    var in_experiment = false;
    var segment = this.get_segment();

    if (this.segment_allocations[segment] !== undefined) {
      var experiment_name = this.segment_allocations[segment];
      var experiment = new this.current_experiments[experiment_name](this.inputs);
      experiment.set_name(`${this.name}-${experiment_name}`);
      experiment.set_salt(`${this.name}-${experiment_name}`);
      this._experiment = experiment;
      this._in_experiment = experiment.in_experiment();
      if (!this._in_experiment) {
        this._assign_default_experiment();
      }
    }
  }

  _assign_default_experiment() {
    this._default_experiment = new this.default_experiment_class(this.inputs);
  }

  default_get(name, default_val) {
    super.require_default_experiment();
    return this._default_experiment.get(name, default_val);
  }

  in_experiment() {
    super.require_experiment();
    return this._in_experiment;
  }

  set_auto_exposure_logging(value) {
    super.require_experiment();
    this._experiment.set_auto_exposure_logging(value);
  }

  get(name, default_val) {
    super.require_experiment();
    if (!this._experiment) {
      return this.default_get(name, default_val);
    } else {
      return this._experiment.get(name, this.default_get(name, default_val));
    }
  }

  log_exposure(extras) {
    super.require_experiment();
    if (!this.experiment) {
      return;
    }
    this._experiment.log_exposure(extras);
  }

  log_event(event_type, extras) {
    super.require_experiment();
    if (!this._experiment) {
      return;
    }
    this._experiment.log_event(event_type, extras);

  }

  //helper function to return the class name of the current experiment class
  getDefaultNamespaceName() {
    if (isObject(this) && this.constructor && this !== this.window) {
      var arr = this.constructor.toString().match(/function\s*(\w+)/);
      if (arr && arr.length === 2) {
        return arr[1];
      }
    }
    return "GenericNamespace";
  }
}

export { Namespace, SimpleNamespace }